### ISSUE TITLE
candle: long-context greedy decode is not bit-reproducible run to run (CPU GEMM reduction order)

### ISSUE THREAD

**Issue by fpt** [fpt/rs-gallium#268]
## Summary

On the native **candle** engine, greedy (`temperature = 0`) decode of the same prompt produces **different token streams across runs** once the context is long (~2000 tokens). It is not a gallium bug — the divergence is in candle-core's own CPU matmul (the `gemm` crate), whose tiled parallel reduction order depends on rayon pool state and so is not bit-reproducible. In f32 that reorder is a ~1 ULP difference per matmul, which compounds over layers/steps until a borderline greedy token tips one way vs the other.

## Repro

`crates/gallium-models/tests/integration.rs`:

- `gpt_oss_gguf_fused_decode_is_deterministic` at long context — override the short defaults:
  ```
  GALLIUM_DEVICE=cpu GALLIUM_KVTEST_FILLER=200 GALLIUM_KVTEST_GEN=48 \
    cargo test --release -p gallium-models --test integration \
    gpt_oss_gguf_fused_decode_is_deterministic -- --ignored --nocapture
  ```
  Two fresh loads of `unsloth/gpt-oss-20b-GGUF` (~2027-token prompt), same greedy params. Observed: **~1 run in 3 diverges** (first token difference around position 25 of 48); the other ~2/3 match. The test's committed defaults (`FILLER=20`, `GEN=24`) stay in the short-context regime where it's stable, on purpose.

- `gpt_oss_gguf_kv_narrowing_is_exact_and_faster` — same effect surfaces as its `assert_eq!(on_ids, off_ids)` ("narrowed K/V must produce the identical greedy stream") failing at large `GALLIUM_KVTEST_FILLER`. `narrow_kv_to_mask` is mathematically exact (softmax over −inf columns contributes exactly 0), so the difference between the narrowed and full paths is purely the GEMM computing 128 score columns vs 2027-then-sliced — different tiling, different reduction order. Passes at `FILLER=60`, flaky at `FILLER=200`.

Not GPT-OSS-specific — any candle model doing long-context CPU decode is exposed. Metal/CUDA not characterised here.

## Impact

- Testsuite: `gpt-oss-20b-candle` / `gpt-oss-120b-candle` run at `temperature = 1.0`, so this is masked there, but any greedy candle run or A/B at long context is non-reproducible.
- Debugging / bisection across candle changes is unreliable at long context.
- It is *not* a correctness bug — every run is a valid greedy decode, they just differ by a late borderline token.

## Things to try

- [ ] **Single-threaded** (`RAYON_NUM_THREADS=1`): if decode is then deterministic, that confirms the reduction-order hypothesis and gives a (slow) reproducible mode. *(verification in progress)*
- [ ] Check whether candle-core / `gemm` has a deterministic-reduction option, or pin `gemm`'s parallelism.
- [ ] Consider a `GALLIUM_DETERMINISTIC` knob that forces `RAYON_NUM_THREADS=1` (or a serial matmul) for reproducible runs.

## Context

Found while revisiting #232/#258 (KV narrowing) with the fused MXFP4 matvec (#265, #267). The fused row-gather itself is an order-preserving pure map and is bit-identical serial vs parallel — it is not the source. See `docs/VERIFICATION_STATUS.md` ("Fused MXFP4 matvec for GPT-OSS decode").

**Comment by fpt** [comment]
## Single-thread result (`RAYON_NUM_THREADS=1`, `FILLER=200` ≈ 2027 tokens, `GEN=32`)

Two distinct effects, and the thread count only fixes one:

| check | multi-thread (default) | single-thread |
|---|---|---|
| `gpt_oss_gguf_fused_decode_is_deterministic` (run A vs run B) | flaky, ~1/3 diverge | **2/2 identical — deterministic** |
| `gpt_oss_gguf_kv_narrowing_is_exact_and_faster` (`on_ids == off_ids`) | fails at `FILLER≥~200` | **still fails — divergence at token 25/32** |

So:

1. **Run-to-run non-reproducibility is thread-scheduling.** Confirmed: `RAYON_NUM_THREADS=1` makes greedy decode of the same prompt reproducible run to run, even at long context. That's the (slow) deterministic mode this issue asked about — prefill went 118s → **488s** and decode 5.7 → **1.3 tok/s** for this prompt, so it's a debugging/repro knob, not a default.

2. **`narrow_kv` vs full K/V is a *separate*, more fundamental thing that 1 thread does *not* fix.** Even single-threaded, `Q @ K_narrowed^T` (128 columns) and `Q @ K_full^T` (2027 columns, then softmax-masked) produce ~1-ULP-different f32 for the overlapping columns — `gemm` picks different blocking / k-accumulation order by matrix shape regardless of parallelism. `narrow_kv_to_mask` is mathematically exact (−inf columns → exactly 0 weight); it is just not bit-exact *through a BLAS GEMM under a shape change*. This means `gpt_oss_gguf_kv_narrowing_is_exact_and_faster`'s `assert_eq!(on_ids, off_ids)` is asserting something no GEMM guarantees, and should be loosened (assert first-N tokens, or step-1 logit closeness) independent of this issue.

### Suggested split

- This issue (#268): the thread-scheduling run-to-run non-determinism. Fix candidates: a `GALLIUM_DETERMINISTIC` knob that pins `RAYON_NUM_THREADS=1` (or a serial matmul), and/or check whether `gemm`/candle expose a deterministic-reduction mode.
- Separate follow-up: loosen the `narrow_kv` exactness assertion (it's a GEMM-under-reshape property, unrelated to threads) — this is also the open call on #232/#258 about whether to keep KV narrowing default-on.
